gems/smithy-cbor/lib/smithy-cbor/deserializer.rb

```diff
@@ -48,43 +48,56 @@ def map(ref, values, target = nil)
         values.each do |key, value|
           next if value.nil? && !sparse?(ref.shape)
 
-          target[key] = value.nil? ? nil : shape(ref.shape.value, value)
+          target[key] = shape(ref.shape.value, value)
         end
         target
       end
 
       def structure(ref, values, target = nil)
-        return Schema::EmptyStructure.new if ref.shape == Prelude::Unit
-
         target = ref.shape.type.new if target.nil?
         ref.shape.members.each do |member_name, member_ref|
-          key = member_ref.member_name
-          next unless values.key?(key)
-
-          target[member_name] = shape(member_ref, values[key])
+          value = values[member_ref.member_name]
+          value = default(member_ref) if value.nil? && default?(member_ref.traits)
+          target[member_name] = shape(member_ref, value)
         end
         target
       end
 
       def union(ref, values, target = nil) # rubocop:disable Metrics/AbcSize
-        raise ArgumentError, "union value includes more than one key, received: #{values.keys}" if values.size > 1
-
-        key, value = values.first
-        return nil if key.nil?
-
         ref.shape.members.each do |member_name, member_ref|
-          name = member_ref.member_name
-          next unless values.key?(name)
+          value = values[member_ref.member_name]
+          next if value.nil?
 
           target = ref.shape.member_type(member_name) if target.nil?
-          return target.new(shape(member_ref, values[name]))
+          return target.new(shape(member_ref, value))
         end
+
+        values.delete('__type')
```

Comment:
Should we do this earlier within the method? Feels like this `__type` doesn't really matter within union context

Reply:
No I believe we have to do it after - because __type could be a jsonName - so we first check all the members, and if we didn't parse anything, delete it, then check for an unknown.

Reply:
I see. CBOR protocol does not support jsonName trait so I was under the impression that we could just ignore `__type` entirely.

Reply:
Oh. Yes, I suppose we could. I just copied what we had for JSON here.